modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java

```diff
@@ -0,0 +1,212 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.metric;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.function.DoubleSupplier;
+import java.util.function.IntSupplier;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.processors.metric.impl.BooleanGauge;
+import org.apache.ignite.internal.processors.metric.impl.DoubleGauge;
+import org.apache.ignite.internal.processors.metric.impl.IntGauge;
+import org.apache.ignite.internal.processors.metric.impl.LongGauge;
+import org.apache.ignite.internal.processors.metric.impl.ObjectGauge;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.spi.metric.Metric;
+import org.apache.ignite.internal.processors.metric.impl.DoubleMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HitRateMetric;
+import org.apache.ignite.internal.processors.metric.impl.IntMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongAdderMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.BooleanMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HistogramMetric;
+import org.apache.ignite.internal.processors.metric.impl.ObjectMetricImpl;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.metricName;
+import static org.apache.ignite.internal.util.lang.GridFunc.nonThrowableSupplier;
+
+/**
+ * Simple implementation.
+ */
+public class MetricRegistryImpl implements MetricRegistry {
+    /** Registered metrics. */
+    private final ConcurrentHashMap<String, Metric> metrics = new ConcurrentHashMap<>();
+
+    /** Logger. */
+    @Nullable private final IgniteLogger log;
+
+    /** Metric set creation listeners. */
+    private final List<Consumer<Metric>> metricCreationLsnrs = new CopyOnWriteArrayList<>();
+
+    /** For test usage only. */
+    public MetricRegistryImpl() {
+        this.log = null;
+    }
+
+    /**
+     * @param log Logger.
+     */
+    public MetricRegistryImpl(IgniteLogger log) {
+        this.log = log;
+    }
+
+    /** {@inheritDoc} */
+    @Override public MetricRegistry withPrefix(String prefix) {
```

Comment:
Calling this method multiple times will produce unnecessary garbage.

Reply:
What do you mean by "unnecessary"?

`withPrefix` method required to make cleaner code for metrics creation.

Reply:
Each call of this method produce new proxy object. I think it's better to change implementation to get rid of new object creation. However, this method is not expected to be called often, so it's up to you to fix it.

Reply:
>  get rid of new object creation

I understand your concern.
I agree we will win some memory here.
But we violate code readability a lot, from my point of view.
I expect that we will construct this object only on metric initialization or other rare operations.

So I will keep it as is.